web/pandas/pdeps/0013-standardize-apply.md

```diff
@@ -0,0 +1,151 @@
+# PDEP-13: Deprecate the apply method on Series & DataFrame and make the agg and transform methods operate on series data
+
+- Created: 24 August 2023
+- Status: Under discussion
+- Discussion: [#52140](https://github.com/pandas-dev/pandas/issues/52509)
+- Author: [Terji Petersen](https://github.com/topper-123)
+- Revision: 2
+
+## Abstract
+
+The `apply`, `transform` and `agg` methods have very complex behavior because they in some cases operate on elements in series, in some cases on series and sometimes try one first, and it that fails, falls back to try the other. There is not a logical system how these behaviors are arranged and it can therefore be difficult for users to understand these methods.
+
+It is proposed that `apply`, `transform` and `agg` in the future will work as follows:
+
+1. the `agg` & `transform` methods of `Series`, `DataFrame` & `groupby` will always operate series-wise and never element-wise
+2. `Series.apply` & `DataFrame.apply` will be deprecated.
+3. `groupby.apply` will not be deprecated (because it behaves differently than `Series.apply` & `DataFrame.apply`)
+
+The above changes means that the future behavior, when users want to apply arbitrary callables in pandas, can be described as follows:
+
+1. When users want to operate on single elements in a `Series` or `DataFrame`, they should use `Series.map` and `DataFrame.map` respectively.
+2. When users want to aggregate a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.agg`, `DataFrame.agg` and `groupby.agg` respectively.
+3. When users want to transform a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.transform`, `DataFrame.transform` and `groupby.transform` respectively.
```

Comment:
I think you need to have the case when users want to have an operation on a row of a `DataFrame`, they should use `DataFrame.agg(func, axis=1)` .  Today a common use case is using `DataFrame.apply(func, axis=1)`.  Since `func` can do anything, it isn't entirely clear that `agg` with `axis=1` will supply each row of the series to the function, meaning that the corresponding function doesn't need to "aggregate".

Reply:
`DataFrame.apply(func, axis=1)` is just `DataFrame.T.apply(func).T`, so any issue with `axis=1` will also be present with `axis=0`.  So I don't see a special issue with `axis=1`, except maybe some idioms that are more common with `axis=1` (which may be a valid concern, though).

Can you give some example code snippets showing some possible problems?

Reply:
@Dr-Irv I've added an example here https://github.com/pandas-dev/pandas/pull/54747/files#r1339884642

Inclined to agree that `.agg` feels really unnatural here